src/service-worker-core.js

```diff
@@ -0,0 +1,202 @@
+}
+
+function basename(url) {
+  return url.substr(url.lastIndexOf('/') + 1);
+}
+
+function versionedUrl(url, version = VERSION) {
+  return url.replace(ampVersion(url), version);
+}
+
+/**
+ * Is this a CDN JS file?
+ * @param {string} url
+ * @return {boolean}
+ */
+function isCdnJsFile(url) {
+  return startsWith(url, 'https://cdn.ampproject.org/') &&
+    endsWith(url, '.js');
+}
+
+
+let cache;
+let db;
+let cacheCleanup = Promise.resolve();
+
+const cachePromise = caches.open('cdn-js').then((result) => {
+  cache = result;
+});
+
+const dbPromise = cachePromise.then(() => {
+  return indexedDBP.open('cdn-js', VERSION, {
+    upgrade(db, { oldVersion, transaction }) {
+      if (oldVersion == 0) {
+        db.createObjectStore('clients', {keyPath: 'clientId'});
+        const files = db.createObjectStore('js-files', {keyPath: 'file'});
+        files.createIndex('versions', 'versions', { multiEntry: true });
+      }
+
+      const files = transaction.objectStore('js-files');
+      const versions = files.index('versions');
+      // Do not serve versions older than two weeks.
+      const cutoff = Number(RELEASE_DATE.setDate(-14));
+      const range = IDBKeyRange.upperBound(cutoff);
+
+      const removals = [];
+      cacheCleanup = versions.openCursor(range).while((cursor) => {
+        const item = cursor.item;
+        const removal = {
+          url: item.url,
+          versions: item.versions.filter((v) => (v <= cutoff))
+        };
+        item.versions = item.versions.filter((v) => v > cutoff);
+
+        return cursor.put(item).then(() => removal);
+      }).then((removals) => {
+        const deletes = removals.map((removal) => {
+          const deletes = removal.versions.forEach((version) => {
+            const url = versionedUrl(removal.url, version);
+            return cache.delete(url);
+          });
+
+          return Promise.all(deletes);
+        });
+
+        return Promise.all(deletes);
+      });
+    }
+  });
+}).then((result) => {
+  db = result;
+});
+
+self.addEventListener('install', function(install) {
+  install.waitUntil(Promise.all([cachePromise, cacheCleanup, dbPromise]));
+
+  // Setup the Fetch listener
+  // My assumptions:
+  //   - The AMP core runtime is the first request (since it's sync).
+  //   - Doc requests one uniform AMP version for all files, anything else
+  //     is malarkey.
+  self.addEventListener('fetch', function(event) {
+    const { request, clientId } = event;
+    const { url } = request;
+    let response;
+
+    // We only cache CDN JS files, and we need a clientId to do our magic.
+    if (clientId && isCdnJsFile(url)) {
+      const requestFile = basename(url);
+      const requestVersion = ampVersion(url);
+
+      // What version do we have for this client? Keep it the same regardless.
+      response = db.transaction('clients', 'readonly').then((transaction) => {
+        const clients = transaction.objectStore('clients');
```

Comment:
Can't since it's tied to a transaction and transaction's only last while there's active work on the `IndexedDB` thread.